fastestimator/estimator.py

```diff
@@ -64,92 +64,104 @@ def __init__(self,
-        self.traces.append(Logger(log_names=self.monitor_names, loss_names=loss_keys))
-
-    def _initialize_loss_keys(self) -> Set[str]:
-        loss_keys = set()
-        for op in self.network.ops:
-            if isinstance(op, UpdateOp):
-                loss_keys = loss_keys.union(set(to_list(op.inputs)))
-        return loss_keys
+        self.trace_inputs = dict()
+        traces = [trace for trace in self.traces]
+        loss_keys = self.network.get_loss_keys()
+        monitor_names = set(filter(None, to_list(self.monitor_names))).union(loss_keys)
+        traces.insert(0, TrainEssential(monitor_names=monitor_names))
+        modes = self.pipeline.get_modes() - {"test"}
+        if "eval" in modes:
+            traces.append(EvalEssential(loss_keys=loss_keys))
+        for mode in modes:
+            trace_inputs = set()
+            for trace in traces:
+                if trace.mode is None or mode in to_list(trace.mode):
+                    trace_inputs.update(filter(None, to_list(trace.inputs)))
+                    monitor_names.update(filter(None, to_list(trace.log_names)))
+            self.trace_inputs[mode] = trace_inputs
+        traces.append(Logger(log_names=monitor_names, loss_names=loss_keys))
+        return traces
 
     def _start(self):
         self._run_traces_on_begin()
         for self.system.epoch_idx in range(self.epochs):
             self.system.mode = "train"
             self._run_epoch()
-            if self.do_eval:
+            if "eval" in self.pipeline.get_modes():
                 self.system.mode = "eval"
                 self._run_epoch()
             self.system.update_epoch_idx()
         self._run_traces_on_end()
 
     def _run_epoch(self):
         self._run_traces_on_epoch_begin()
-        self.system.batch_size = self.pipeline.get_batch_size(mode=self.system.mode, epoch=self.system.epoch_idx)
-        ds_iter = self.pipeline.get_iterator(mode=self.system.mode, epoch=self.system.epoch_idx)
-        for self.system.batch_idx, batch in enumerate(ds_iter):
-            if self.network.framework == "tensorflow":
-                batch = torch_to_tf(batch)  # TODO - this should maybe be handled somewhere else...
+        self.network.load_epoch(mode=self.system.mode, epoch=self.system.epoch_idx)
+        self.system.loader = self.pipeline.get_loader(mode=self.system.mode, epoch=self.system.epoch_idx)
+        for self.system.batch_idx, batch in enumerate(self.system.loader):
             if self.system.batch_idx == self.steps_per_epoch and self.system.mode == "train":
                 break
+            batch = self._configure_tensor(batch)
             self._run_traces_on_batch_begin()
-            prediction = self.network.run_step(batch, {"mode": self.system.mode, "epoch": self.system.epoch_idx})
+            prediction = self.network.run_step(batch, {"mode": self.system.mode, "warmup": False})
```

Comment:
note to self: look at these args on second pass (why dict?)

Reply:
after looking things over, not clear to me why we don't just put system inside network rather than these dictionary args. is it to do with tf.function problems?

Reply:
yes, tf.function has issue accessing class variable